torch/distributed/CONTRIBUTING.md

```diff
@@ -0,0 +1,69 @@
+# Contributing to PyTorch Distributed
+
+Please go through PyTorch's top level [Contributing Guide](../../CONTRIBUTING.md) before proceeding with this guide.
+
+[PyTorch Distributed Overview](https://pytorch.org/tutorials//beginner/dist_overview.html) is a great starting point with a lot of tutorials, documentation and design docs covering PyTorch Distributed. We would highly recommend going through some of that material before you start working on PyTorch Distributed.
+
+In this document, we mostly focus on some of the code structure for PyTorch distributed and implementation details.
+
+## C10D and DistributedDataParallel
+
+The figure below demonstrates building blocks of the c10d and DDP package and shows how typically an application is layered on top. Most parts of the distributed package are implemented in C++ and then bound to the Python frontend (see [c10d/init.cpp](../csrc/distributed/c10d/init.cpp)).
+
+![C10D_ARCH](../../docs/source/_static/img/pt_distributed_arch.png)
+
+### Process Groups
+
+Process groups (PG) take care of communications across processes. It is up to users to decide how to place processes, e.g., on the same machine or across machines. PG exposes a set of communication APIs, e.g., send, recv, broadcast, allgather, allreduce, etc.
+
+Source Code: [ProcessGroup.cpp](../lib/c10d/ProcessGroup.cpp) and  [ProcessGroup.hpp](../lib/c10d/ProcessGroup.hpp)
+
+#### Process Group Backends
+
+We currently offer three backends for Process Groups: [ProcessGroupGloo.hpp](../lib/c10d/ProcessGroupGloo.hpp), [ProcessGroupMPI.hpp](../lib/c10d/ProcessGroupMPI.hpp) and [ProcessGroupNCCL.hpp](../torch/lib/c10d/ProcessGroupNCCL.hpp)
+
+#### Store
+
+Processes discover each other through a rendezvous process on a common Store (See [Store.hpp](../torch/lib/c10d/Store.hpp) for the interface and [FileStore.hpp](../torch/lib/c10d/FileStore.hpp), [TCPStore.hpp](../torch/lib/c10d/TCPStore.hpp) and [PrefixStore.hpp](../torch/lib/c10d/PrefixStore.hpp) for implementations.)
+
+### Distributed Data Parallel
+
+DDP is implemented as a module in [distributed.py](../nn/parallel/distributed.py) with some of the core functions implemented in [reducer.cpp](../csrc/distributed/c10d/reducer.cpp) and [comm.cpp](../csrc/distributed/c10d/reducer.cpp). Gradients synchronizations occur in backward pass, triggered as autograd hooks.
```

Comment:
Should we link to DDP design (https://pytorch.org/docs/stable/notes/ddp.html) and maybe do the same for RPC / Dist Autograd? I guess we may not need to since these links are available from the dist_overview page.